hazelcast/src/main/java/com/hazelcast/query/Predicates.java

```diff
@@ -29,77 +29,369 @@
 import com.hazelcast.query.impl.predicates.OrPredicate;
 import com.hazelcast.query.impl.predicates.RegexPredicate;
 
+import java.util.Date;
+
 /**
  * A utility class to create {@link com.hazelcast.query.Predicate} instances.
+ * <p>
+ * <b>Special Attributes</b>
+ * <p>
+ * The predicate factory methods accepting an attribute name support two special attributes:
+ * <ul>
+ * <li>{@link QueryConstants#KEY_ATTRIBUTE_NAME "__key"} - instructs the predicate to act on the key associated
+ * with an item.
+ * <li>{@link QueryConstants#THIS_ATTRIBUTE_NAME "this"} - instructs the predicate to act on the value associated
+ * with an item.
+ * </ul>
+ * <p>
+ * <b>Attribute Paths</b>
+ * <p>
+ * Dot notation may be used for attribute name to instruct the predicate to act on the attribute located at deeper
+ * level of an item: given {@code "fullName.firstName"} path the predicate will act on {@code firstName} attribute
+ * of the value fetched by {@code fullName} attribute from the item itself.
+ * <p>
+ * Square brackets notation may be used to instruct the predicate to act on the array/collection element at the
+ * specified index: given {@code "names[0]"} path the predicate will act on the first item of the array/collection
+ * fetched by {@code names} attribute from the item.
```

Comment:
I'm curious what would happen if I ask for `names[10]` and collection only has 5 elements. Also when asking for `person.firstName` and the `person` is `null`.

Reply:
Done. I think we have to stop at this point ;)

Reply:
Good that you documented what happens when the attribute does not exist. It would be better if 
```
@throws IllegalArgumentException If the attribute does not exist
```
 is present at each method.